Improve launch time of WebKit processes

pvollan opened the pull request:
#### c61ab1f570586f0ceaab0f4eb690ea1718940eca
<pre>
Improve launch time of WebKit processes
<a href="https://bugs.webkit.org/show_bug.cgi?id=260198">https://bugs.webkit.org/show_bug.cgi?id=260198</a>
rdar://113901020

Reviewed by Brent Fulgham.

Experiments show that it is signficantly faster to launch WebKit processes as extensions with ExtensionKit.
This patch is working towards this by creating empty executables for these extensions and starting
adopting ExtensionKit (<a href="https://developer.apple.com/documentation/extensionkit)">https://developer.apple.com/documentation/extensionkit)</a> as the long-term architecture
for WebKit auxiliary processes as a lighter-weight mechanism for launching and handling processes.

This patch is written by Alex Christensen and Per Arne Vollan.

* Configurations/CommonBase.xcconfig:
* Source/WebKit/Configurations/BaseExtension.xcconfig: Added.
* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKit/Configurations/GPUExtension.xcconfig: Added.
* Source/WebKit/Configurations/NetworkingExtension.xcconfig: Added.
* Source/WebKit/Configurations/WebContentCaptivePortalExtension.xcconfig: Added.
* Source/WebKit/Configurations/WebContentCrashyExtension.xcconfig: Added.
* Source/WebKit/Configurations/WebContentExtension.xcconfig: Added.
* Source/WebKit/Shared/API/Cocoa/WKMain.h:
* Source/WebKit/Shared/API/Cocoa/WKMain.mm:
(WKExtensionMain):
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.entitlements: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.swift: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.h: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.mm: Added.
(extensionMain):
* Source/WebKit/Shared/AuxiliaryProcessExtensions/GPUExtension-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/NetworkingExtension-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-CaptivePortal-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-Crashy-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/com.apple.WebKit.appexpt: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/268228@main">https://commits.webkit.org/268228@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c9374784f9546703090d0a6a8f93a2bd96d021a6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19045 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19389 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19997 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20911 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17804 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22682 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19554 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19559 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19268 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19368 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16569 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21793 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16576 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23746 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17617 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17526 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21693 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18101 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17190 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4544 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21547 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17936 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -17706,15 +18110,22 @@
 				PRODUCT_NAME = All;
+				SWIFT_OBJC_BRIDGING_HEADER = "Shared/AuxiliaryProcessExtensions/Framework, XPC Services, Extensions, and daemons-Bridging-Header.h";
+				SWIFT_VERSION = 5.0;
 			};
```
Be sure to remove these, and put all settings overrides in xcconfigs. Xcode adds some settings here by default when you create a new target, but we don't want them.

Ditto for the other `XCBuildConfiguration` objects below this.

pvollan replied:
Fixed!

Thanks for reviewing!

emw-apple commented on Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.mm:
```diff
@@ -0,0 +1,33 @@
+
+void extensionMain(int argc, char** argv)
+{
+    WKExtensionMain(argc, argv);
```
nit: we should probably handle the return value here, right?
```suggestion
    exit(WKExtensionMain(argc, argv));
```

pvollan replied:
Good point, I have not addressed this yet, but will do.

emw-apple commented on Source/WebKit/Configurations/BaseExtension.xcconfig:
```diff
@@ -0,0 +1,32 @@
+SWIFT_OBJC_BRIDGING_HEADER = Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.h;
+PRODUCT_NAME = $(TARGET_NAME);
+CLANG_USE_OPTIMIZATION_PROFILE = NO;
+CLANG_INSTRUMENT_FOR_OPTIMIZATION_PROFILING_ENABLE_LLVM_PROFILE_GENERATION = NO;
```
I assume the intent is to disable PGO instrumentation entirely? You can just set `CLANG_INSTRUMENT_FOR_OPTIMIZATION_PROFILING = NO`.

pvollan replied:
Fixed.

emw-apple reviewed (commented):
Nice! Let me know if I can be of help testing this.

xeenon commented on Source/WebKit/Shared/AuxiliaryProcessExtensions/com.apple.WebKit.appexpt:
```diff
@@ -0,0 +1,46 @@
+    <key>com.apple.webkit.webcontent.extension</key>
+    <dict>
+        <key>EXExtensionPointIsPublic</key>
+        <true/>
```
I think you want all these `EXExtensionPointIsPublic` values to be `<false/>`. These are not extension points that third-party apps can implement extensions for.

pvollan replied:
Fixed!

Thanks for reviewing, weinig, xeenon, and mew-apple!

brentfulgham commented on Configurations/CommonBase.xcconfig:
```diff
@@ -55,6 +55,13 @@ OTHER_LDFLAGS = $(inherited) $(WK_COMMON_OTHER_LDFLAGS);
+SWIFT_OPTIMIZATION_LEVEL = $(WK_DEFAULT_SWIFT_OPTIMIZATION_LEVEL);
+
+WK_DEFAULT_SWIFT_VERSION = 5.0;
+SWIFT_VERSION = $(WK_DEFAULT_SWIFT_VERSION);
```
I'm a bit surprised that these rules aren't visible to `BaseTarget.xcconfig` or `BaseExtension.xcconfig`. I wonder if we should have a Swift xcconfig file (not necessary for this patch).

pvollan replied:
That is a good point. I think the changes here may not be required. Will fix.

Thanks for reviewing!

brentfulgham reviewed (approved):
Assuming you incorporate Elliott's suggested change, I think this looks great.

pvollan committed:
Improve launch time of WebKit processes
https://bugs.webkit.org/show_bug.cgi?id=260198
rdar://113901020

Reviewed by Brent Fulgham.

Experiments show that it is signficantly faster to launch WebKit processes as extensions with ExtensionKit.
This patch is working towards this by creating empty executables for these extensions and starting
adopting ExtensionKit (https://developer.apple.com/documentation/extensionkit) as the long-term architecture
for WebKit auxiliary processes as a lighter-weight mechanism for launching and handling processes.

This patch is written by Alex Christensen and Per Arne Vollan.

* Configurations/CommonBase.xcconfig:
* Source/WebKit/Configurations/BaseExtension.xcconfig: Added.
* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKit/Configurations/GPUExtension.xcconfig: Added.
* Source/WebKit/Configurations/NetworkingExtension.xcconfig: Added.
* Source/WebKit/Configurations/WebContentCaptivePortalExtension.xcconfig: Added.
* Source/WebKit/Configurations/WebContentCrashyExtension.xcconfig: Added.
* Source/WebKit/Configurations/WebContentExtension.xcconfig: Added.
* Source/WebKit/Shared/API/Cocoa/WKMain.h:
* Source/WebKit/Shared/API/Cocoa/WKMain.mm:
(WKExtensionMain):
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.entitlements: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.swift: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.h: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.mm: Added.
(extensionMain):
* Source/WebKit/Shared/AuxiliaryProcessExtensions/GPUExtension-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/NetworkingExtension-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-CaptivePortal-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-Crashy-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/com.apple.WebKit.appexpt: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/268228@main
